### PR TITLE
Feature/form threats

### PR DESCRIPTION
**Pull request by onyxvd**
Create Threats form
* [x] Create new form converter (refer to  BirdsConverter)
* [x] Create new action in SmartBirdsApi for submitting the form
* [x] Create new form uploader (refer to  BirdsUploader)
* [x] Create new form layout (refer to fragment_monitoring_form_new_ciconia_entry.xml)
* [x] Create new form entry fragment (refer to NewCiconiaEntryFormFragment)
* [x] Add new item in res/menu/monitoring.xml
* [x] Add new item in EntryType enum and new item in MENU_ACTION_IDS array
* [x] Add new item in @OptionsItem for setFormType() in MonitoringActivity
* [x] Update bind() in MonitoringEntryListRowPartialView
* [x] Update nomenclatures.json and species.json

Introduce SingleChoiceConfigFormInput
Introduce SingleChoiceRadioFormInput

### CODE REVIEW

File: app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/NewThreatsEntryRequiredFormFragment.java [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        }
+
+        switch (type) {
+            case "dead":
+                showDeadForm();
+                break;
+            case "alive":
+                showAliveForm();
+                break;
+            case "bait":
```
Comment (groupsky) [inline app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/NewThreatsEntryRequiredFormFragment.java]: again strings should be extracted as constants

File: app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/NewThreatsEntryRequiredFormFragment.java [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        hideAllFields();
+        if (primaryTypeValue == null) {
+            return;
+        }
+
+        switch (primaryTypeValue) {
+            case "threat":
+                showThreatForm();
+                break;
+            case "poison":
```
Comment (groupsky) [inline app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/NewThreatsEntryRequiredFormFragment.java]: extract to constants

File: app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/NewThreatsEntryRequiredFormFragment.java [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+            case "herptiles":
+                species.setKey("species_herptiles");
+                break;
+            case "mammals":
+                species.setKey("species_mammals");
+                break;
+            case "invertebrates":
+                species.setKey("species_invertebrates");
+                break;
+            case "plants":
```
Comment (groupsky) [inline app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/NewThreatsEntryRequiredFormFragment.java]: constants or at least a hashmap

File: app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/NewThreatsEntryRequiredFormFragment.java [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+            }
+
+            switch (poisonedTypeValue) {
+                case "dead":
+                    showDeadForm();
+                    break;
+                case "alive":
+                    showAliveForm();
+                    break;
+                case "bait":
```
Comment (groupsky) [inline app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/NewThreatsEntryRequiredFormFragment.java]: missing constants

File: app/src/main/java/org/bspb/smartbirds/pro/ui/partial/MonitoringEntryListRowPartialView.java [hunk 5]
```diff
@@ -131,6 +133,22 @@ public void bind(MonitoringEntry entry) {
                 speciesView.setText(entry.data.get(context.getString(R.string.tag_species_scientific_name)));
                 countView.setText(entry.data.get(context.getString(R.string.tag_count)));
                 break;
+            case THREATS:
+                String primaryType = entry.data.get(context.getString(R.string.tag_primary_type));
+
+                if ("threat".equalsIgnoreCase(primaryType)) {
```
Comment (groupsky) [inline app/src/main/java/org/bspb/smartbirds/pro/ui/partial/MonitoringEntryListRowPartialView.java]: extract to constant

File: app/src/main/java/org/bspb/smartbirds/pro/ui/partial/MonitoringEntryListRowPartialView.java [hunk 6]
```diff
@@ -131,6 +133,22 @@ public void bind(MonitoringEntry entry) {
...
+                    int idx = Arrays.asList(FormsConfig.threats_primary_types.getValues()).indexOf(primaryType);
+                    typeView.setText(FormsConfig.threats_primary_types.getLabels()[idx]);
```
Comment (groupsky) [inline app/src/main/java/org/bspb/smartbirds/pro/ui/partial/MonitoringEntryListRowPartialView.java]: Can this be extracted into a helper method?

File: app/src/main/java/org/bspb/smartbirds/pro/ui/partial/MonitoringEntryListRowPartialView.java [hunk 7]
```diff
@@ -131,6 +133,22 @@ public void bind(MonitoringEntry entry) {
...
+                    speciesView.setText(entry.data.get(context.getString(R.string.tag_category)));
+                } else {
+                    String poisonedType = entry.data.get(context.getString(R.string.tag_poisoned_type));
+                    int idxPoisoned = Arrays.asList(FormsConfig.threats_poisoned_types.getValues()).indexOf(poisonedType);
+                    int idxPrimary = Arrays.asList(FormsConfig.threats_primary_types.getValues()).indexOf(primaryType);
+                    typeView.setText(FormsConfig.threats_primary_types.getLabels()[idxPrimary]);
+                    speciesView.setText(FormsConfig.threats_poisoned_types.getLabels()[idxPoisoned]);
```
Comment (groupsky) [inline app/src/main/java/org/bspb/smartbirds/pro/ui/partial/MonitoringEntryListRowPartialView.java]: same here - can the conversion be extracted into a helper method?

File: app/src/main/java/org/bspb/smartbirds/pro/ui/utils/FormsConfig.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            new int[]{R.string.class_birds, R.string.class_herptiles, R.string.class_mammals, R.string.class_invertebrates, R.string.class_plants}
+    ),
+    threats_primary_types(
+            new String[]{"threat", "poison"},
+            new int[]{R.string.primary_type_threat, R.string.primary_type_poison}
+    ),
+    threats_poisoned_types(
+            new String[]{"dead", "alive", "bait"},
+            new int[]{R.string.poisoned_type_dead, R.string.poisoned_type_alive, R.string.poisoned_type_bait}
+    );
```
Comment (groupsky) [inline app/src/main/java/org/bspb/smartbirds/pro/ui/utils/FormsConfig.java]: Seems there are enums that contain the directly used string - can extract the strings as constants or have separate enums for each config that can provide the value and string.id for each item

File: app/src/main/java/org/bspb/smartbirds/pro/ui/utils/FormsConfig.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        }
+    }
+
+    public static List<NomenclatureConfig[]> configs = new ArrayList<NomenclatureConfig[]>() {
+        {
+            add(SpeciesClass.values());
+            add(ThreatsPrymaryType.values());
+            add(ThreatsPoisonedType.values());
+        }
+    };
```
Comment (groupsky) [inline app/src/main/java/org/bspb/smartbirds/pro/ui/utils/FormsConfig.java]:
```suggestion
    );
```

File: app/src/main/java/org/bspb/smartbirds/pro/ui/views/SingleChoiceConfigFormInput.java [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+                    break;
+                default:
+                    mLastSelected = mAdapter.getItem(which);
+                    onClick(dialog, BUTTON_POSITIVE);
+                    break;
+            }
+        }
+    }
+
+    class ConfigItem {
```
Comment (groupsky) [inline app/src/main/java/org/bspb/smartbirds/pro/ui/views/SingleChoiceConfigFormInput.java]: Why not use directly `FormsConfig.NomenclatureConfig`?

File: app/src/main/res/values/attrs.xml [hunk 11]
```diff
@@ -23,6 +24,12 @@
 
     <attr name="entries" format="string" />
     <attr name="hint" format="string" />
+    <!--Value of the enum represent configs index in FormsConfig-->
```
Comment (groupsky) [inline app/src/main/res/values/attrs.xml]:
```suggestion
    <!-- Value of the enum represent configs index in FormsConfig -->
```

File: app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/NewThreatsEntryRequiredFormFragment.java [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+
+    private void handlePoisonedType(String type) {
+        hideAllFields();
+        poisonedType.setVisibility(View.VISIBLE);
+
+        if (type == null) {
+            return;
+        }
+
+        FormsConfig.ThreatsPoisonedType poisonedType = FormsConfig.ThreatsPoisonedType.valueOf(type);
```
[the remark this answers is not in the document]
Comment (onyxvd) [reply app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/NewThreatsEntryRequiredFormFragment.java]: There is no need for case insensitive matching. Refactored the rest of code to use exact match.

**Pull request merged**